File: 84.largest-rectangle-in-histogram.py
class Solution:
    def largestRectangleArea(self, heights):
        """
        :type heights: List[int]
        :rtype: int
        """
        ans = 0
        index = []
        for i in range(len(heights) + 1):
            while index and (i == len(heights) or heights[index[-1]] > heights[i]):
                h = heights[index.pop()]
                if not index:
                    ans = max(ans, h * i)
                else:
                    ans = max(ans, h * (i - index[-1] - 1))
            index.append(i)
        return ans;

File: 84.largest-rectangle-in-histogram.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        int ans = 0;
        stack<int> index;
        for (int i = 0; i <= heights.size(); i++) {
            while (!index.empty() && (i == heights.size() || heights[index.top()] > heights[i])) {
                int h = heights[index.top()];
                index.pop();
                if (index.empty()) {
                    ans = max(ans, h * i);
                } else {
                    ans = max(ans, h * (i - index.top() - 1));
                }
            }
            index.push(i);
        }
        return ans;
    }
};